onyx_otc/websocket_v2.py

```diff
@@ -0,0 +1,249 @@
+def on_exit(cli: OnyxWebsocketClientV2) -> None:
+    logger.info("Connection closed")
+
+
+@dataclass
+class OnyxWebsocketClientV2:
+    """
+    WebSocket client for the Onyx OTC API v2.
+
+    Attributes:
+        api_token: API authentication token
+        ws_url: WebSocket endpoint URL
+        on_response: Callback for handling responses
+        on_event: Callback for handling channel events
+        on_exit: Callback for handling connection closure
+    """
+
+    api_token: str = field(default_factory=lambda: os.environ.get("ONYX_API_TOKEN", ""))
+    ws_url: str = field(
+        default_factory=lambda: os.environ.get(
+            "ONYX_WS_V2_URL", "wss://ws.dev.onyxhub.co/stream/v2/binary"
+        )
+    )
+
+    on_response: ResponseHandler = field(default=on_response)
+    on_event: EventHandler = field(default=on_event)
+    on_exit: ExitHandler = field(default=on_exit)
+
+    _queue: asyncio.Queue[OtcRequest] = field(default_factory=asyncio.Queue)
+    _ws: ClientWebSocketResponse | None = field(default=None, init=False)
+    _write_task: asyncio.Task | None = field(default=None, init=False)
+    _is_running: bool = field(default=False, init=False)
+    _reconnect_delay: float = field(default=1.0, init=False)
+    _max_reconnect_delay: float = field(default=60.0, init=False)
+
+    def _create_request(self, method: Method.ValueType, **kwargs: Any) -> OtcRequest:
+        """Create a new request with the current timestamp."""
+        timestamp = Timestamp()
+        timestamp.FromDatetime(datetime.now(UTC))
+        return OtcRequest(method=method, timestamp=timestamp, **kwargs)
+
+    async def _handle_binary_message(self, data: bytes) -> None:
+        """Handle incoming binary messages."""
+        if response := self.parse_response(data):
+            self.on_response(self, response)
+        elif message := self.parse_channel_message(data):
+            self.on_event(self, message)
+        else:
+            logger.warning("Unknown message type received")
+
+    def parse_response(self, data: bytes) -> OtcResponse | None:
+        """Parse binary data as OtcResponse."""
+        try:
+            response = OtcResponse()
+            response.ParseFromString(data)
+
+            if (
+                response.HasField("error")
+                or response.HasField("auth")
+                or response.HasField("subscription")
+                or response.HasField("order")
+            ):
+                return response
+            return None
+        except Exception:
+            logger.debug("Failed to parse as OtcResponse", exc_info=True)
+            return None
+
+    def parse_channel_message(self, data: bytes) -> ChannelMessage | None:
+        """Parse binary data as ChannelMessage."""
+        try:
+            message = ChannelMessage()
+            message.ParseFromString(data)
+            return message
+        except Exception:
+            logger.debug("Failed to parse as ChannelMessage", exc_info=True)
```

Comment:
I dont think we should catch here. Just crash

Reply:
Done